spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindableProxyFactory.java

```diff
@@ -0,0 +1,316 @@
+							input, method);
+					Class<?> inputChannelType = method.getReturnType();
+					MessageChannel sharedChannel = locateSharedChannel(name);
+					if (sharedChannel == null) {
+						MessageChannel inputChannel = createMessageChannel(inputChannelType);
+						inputs.put(name, new ChannelHolder(inputChannel, true));
+					}
+					else {
+						if (inputChannelType.isAssignableFrom(sharedChannel.getClass())) {
+							inputs.put(name, new ChannelHolder(sharedChannel, false));
+						}
+						else {
+							// handle the special case where the shared channel is of a different nature
+							// (i.e. pollable vs subscribable) than the target channel
+							final MessageChannel inputChannel = createMessageChannel(inputChannelType);
+							if (isPollable(sharedChannel.getClass())) {
+								bridgePollableToSubscribableChannel(sharedChannel,
+										inputChannel);
+							}
+							else {
+								bridgeSubscribableToPollableChannel(
+										(SubscribableChannel) sharedChannel, inputChannel);
+							}
+							inputs.put(name, new ChannelHolder(inputChannel, false));
+						}
+					}
+				}
+
+				Output output = AnnotationUtils.findAnnotation(method, Output.class);
+				if (output != null) {
+					String name = BindingBeanDefinitionRegistryUtils.getChannelName(
+							output, method);
+					Class<?> messageChannelType = method.getReturnType();
+					MessageChannel sharedChannel = locateSharedChannel(name);
+					if (sharedChannel == null) {
+						MessageChannel outputChannel = createMessageChannel(messageChannelType);
+						outputs.put(name, new ChannelHolder(outputChannel, true));
+					}
+					else {
+						if (messageChannelType.isAssignableFrom(sharedChannel.getClass())) {
+							outputs.put(name, new ChannelHolder(sharedChannel, false));
+						}
+						else {
+							// handle the special case where the shared channel is of a different nature
+							// (i.e. pollable vs subscribable) than the target channel
+							final MessageChannel outputChannel = createMessageChannel(messageChannelType);
+							if (isPollable(messageChannelType)) {
+								bridgePollableToSubscribableChannel(outputChannel,
+										sharedChannel);
+							}
+							else {
+								bridgeSubscribableToPollableChannel(
+										(SubscribableChannel) outputChannel,
+										sharedChannel);
+							}
+							outputs.put(name, new ChannelHolder(outputChannel, false));
+						}
+					}
+				}
+			}
+		});
+	}
+
+	private MessageChannel locateSharedChannel(String name) {
+		return sharedChannelRegistry != null ? sharedChannelRegistry
+				.getSharedChannel(getNamespacePrefixedChannelName(name)) : null;
+	}
+
+	private String getNamespacePrefixedChannelName(String name) {
+		return channelNamespace + "." + name;
+	}
+
+	private void bridgeSubscribableToPollableChannel(SubscribableChannel sharedChannel,
+			MessageChannel inputChannel) {
+		sharedChannel.subscribe(new MessageChannelBinderSupport.DirectHandler(
```

Comment:
Can we create the same  `DirectHandler` in this class? may be that's in the next step when we remove the local binder :-)